src/ClassDiscovery.php

```diff
@@ -0,0 +1,58 @@
+<?php
+
+namespace Http\Discovery;
+
+/**
+ * Registry that based find results on class existence
+ *
+ * @author David de Boer <david@ddeboer.nl>
+ */
+abstract class ClassDiscovery implements Discovery
+{
+    protected static $cache;
+    protected static $factories = [];
+
+    /**
+     * Register a factory
+     *
+     * @param string $name
+     * @param string $class
+     * @param string $factory
+     */
+    public static function register($name, $class, $factory = null)
+    {
+        static::$cache = null;
+
+        static::$factories[$name] = [
+            'class'   => $class,
+            'factory' => $factory,
+        ];
+    }
+
+    /**
+     * Finds a Factory
+     *
+     * @return object
+     *
+     * @throws NotFoundException
+     */
+    public static function find()
+    {
+        // We have a cache
+        if (isset(static::$cache)) {
+            return static::$cache;
+        }
+
+        foreach (static::$factories as $name => $definition) {
+            if (class_exists($definition['class'])) {
+                if (isset($definition['factory'])) {
```

Comment:
I would rather call call class the one which should be instantiated and condition which should be checked. This way the target class is consistent between all guessers.